File: app/src/main/java/com/raival/fileexplorer/tabs/file/tasks/CopyTask.java
package com.raival.fileexplorer.tabs.file.tasks;

import android.os.Handler;
import android.os.Looper;

import com.raival.fileexplorer.tabs.file.model.Task;
import com.raival.fileexplorer.utils.FileUtil;

import java.io.File;
import java.util.ArrayList;

public class CopyTask extends Task {
    private final ArrayList<File> filesToCopy;
    private File activeDirectory;

    public CopyTask(ArrayList<File> filesToCopy) {
        this.filesToCopy = filesToCopy;
    }

    public ArrayList<File> getFilesToCopy() {
        return filesToCopy;
    }

    @Override
    public String getName() {
        return "Copy";
    }

    @Override
    public String getDetails() {
        StringBuilder sb = new StringBuilder();
        boolean first = true;

        for (File file : filesToCopy) {
            if (!first) {
                sb.append(", ");
            }
            sb.append(file.getName());
            first = false;
        }
        return sb.toString();
    }

    @Override
    public boolean isValid() {
        for (File file : filesToCopy) {
            if (!file.exists())
                return false;
        }
        return true;
    }

    @Override
    public void setActiveDirectory(File directory) {
        activeDirectory = directory;
    }

    @Override
    public void start(OnUpdateListener onUpdateListener, OnFinishListener onFinishListener) {
        new Thread(() -> {
            boolean error = false;
            try {
                int progress = 1;
                for(File file : filesToCopy){
                    try{
                        int finalProgress = progress;
                        new Handler(Looper.getMainLooper()).post(() -> {
                            onUpdateListener.onUpdate("["
                                    + finalProgress
                                    + "/"
                                    + filesToCopy.size()
                                    + "]"
                                    + "Copying "
                                    + file.getName());
                        });
                        FileUtil.copy(file, activeDirectory);
                    } catch (Exception exception){
                        error = true;
                    }
                    ++progress;
                }
                boolean finalError = error;
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish(finalError
                        ? "An error occurred, some files didn't get copied"
                        : "Files copied successfully"));
            } catch (Exception e) {
                e.printStackTrace();
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish(e.toString()));
            }
        }).start();
    }
}

File: app/src/main/java/com/raival/fileexplorer/tab/file/task/CopyTask.kt
package com.raival.fileexplorer.tab.file.task

import com.raival.fileexplorer.tab.file.misc.FileUtils
import com.raival.fileexplorer.tab.file.model.Task
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class CopyTask(private val filesToCopy: ArrayList<File>) : Task() {
    private var activeDirectory: File? = null
    override val name: String
        get() = "Copy"
    override val details: String
        get() {
            val sb = StringBuilder()
            var first = true
            for (file in filesToCopy) {
                if (!first) {
                    sb.append(", ")
                }
                sb.append(file.name)
                first = false
            }
            return sb.toString()
        }
    override val isValid: Boolean
        get() {
            for (file in filesToCopy) {
                if (!file.exists()) return false
            }
            return true
        }

    override fun setActiveDirectory(directory: File) {
        activeDirectory = directory
    }

    override fun start(onUpdate: OnUpdateListener, onFinish: OnFinishListener) {
        CoroutineScope(Dispatchers.IO).launch {
            var error = false
            try {
                var progress = 1
                for (file in filesToCopy) {
                    try {
                        val finalProgress = progress
                        withContext(Dispatchers.Main) {
                            onUpdate.onUpdate(
                                "["
                                        + finalProgress
                                        + "/"
                                        + filesToCopy.size
                                        + "]"
                                        + "Copying "
                                        + file.name
                            )
                        }
                        FileUtils.copy(file, activeDirectory!!, true)
                    } catch (exception: Exception) {
                        error = true
                    }
                    ++progress
                }
                val finalError = error
                withContext(Dispatchers.Main) { onFinish.onFinish(if (finalError) "An error occurred, some files didn't get copied" else "Files copied successfully") }
            } catch (e: Exception) {
                e.printStackTrace()
                withContext(Dispatchers.Main) { onFinish.onFinish(e.toString()) }
            }
        }
    }
}